### PR TITLE
Feature/sof 4397 - Making Unittests Application Flavors

### PR DESCRIPTION
**Pull request by bsmith24**
The unit tests are working. They rely on using pickle files stored in fixtures.

The setup method in each of the BaseUnitTest classes is similar to the one used in the integration tests. At some point in the near future, we could lump these together in a single place, probably also in fixtures.

### CODE REVIEW

File: test/pythonml/test_context.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            for line in unmodified_settings_file:
+                line = re.sub("PROBLEM_CATEGORY_HERE", 'regression', line)
+                file.write(line)
+        import settings
+        self.context = settings.Context()
+        self.context.context_paths.update({self.context._context_file: 'file'})
+
+    def tearDown(self):
+        os.system('rm -rf .job_context')
+        os.system('rm settings.py')
```
Comment (AcylSilane) [inline test/pythonml/test_context.py:27]: Prefer `subprocess.call` over `os.system` 

docs: https://docs.python.org/3/library/subprocess.html#subprocess.call

Example would be `subprocess.call(['rm', '-rf', '.jobcontext'])`

File: test/pythonml/test_context.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class TestContext(unittest.TestCase):
+    """
+    Unit tests for the methods in the Context class defined in the settings.py module
+    """
+
+    def setUp(self):
+        with open('fixtures/settings.py', 'r') as file:
+            unmodified_settings_file = file.readlines()
```
Comment (AcylSilane) [inline test/pythonml/test_context.py:16]: I'd recommend calling this `settings_template` to be a little more explicit about what it is

File: test/pythonml/test_context.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        with open('settings.py', 'w') as file:
```
Comment (AcylSilane) [inline test/pythonml/test_context.py:17]: You can collapse both of these with-statements into a single one. For example:

```python
with open('fixtures/settings.py', 'r') as unmodified_settings_file, open('settings.py', w') as rendered_settings_file:
   ...
```

File: test/pythonml/test_context.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        self.context = settings.Context()
+        self.context.context_paths.update({self.context._context_file: 'file'})
+
+    def tearDown(self):
+        os.system('rm -rf .job_context')
+        os.system('rm settings.py')
+
+    @mock.patch('builtins.open', new_callable=mock.mock_open)
+    @mock.patch('pickle.dump')
+    def test_update_context(self, mock_pickle, mock_builtin_open):
```
Comment (AcylSilane) [inline test/pythonml/test_context.py:31]: I recommend adding a short description to each test describing what it's actually supposed to test. What is the expected result when this test runs?

As an alternative, you could make the test names more descriptive (for example, "test_pickle_dumped_on_context_update" or something)

File: test/pythonml/test_model_flavors.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python
+import importlib
+import os
+import shutil
+import unittest
+from parameterized import parameterized
+from fixtures.unittest_baseclass import BaseUnitTest
```
Comment (AcylSilane) [inline test/pythonml/test_model_flavors.py:7]: The base-class isn't really a fixture, we should move it out of the fixtures folder

File: test/pythonml/test_model_flavors.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        import settings
+        importlib.reload(settings)
+        with settings.context as context:
+            if category == 'regression':
+                rmse = context.load('RMSE')
+                assert(rmse <= 2 * 10)
+            elif category == 'classification':
+                confusion_matrix = context.load('confusion_matrix')
+                accuracy = confusion_matrix.diagonal() / confusion_matrix.sum(axis=0)
+                assert (accuracy.all() >= 0.6)
```
Comment (AcylSilane) [inline test/pythonml/test_model_flavors.py:26]: Extract the magic numbers for the RMSE and Accuracy thresholds to variables.

File: test/pythonml/test_model_flavors.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    This class performs unit tests for the model flavors
+    """
+
+    def check_pass_conditions_training(self, category):
+        import settings
+        importlib.reload(settings)
+        with settings.context as context:
+            if category == 'regression':
+                rmse = context.load('RMSE')
+                assert(rmse <= 2 * 10)
```
Comment (AcylSilane) [inline test/pythonml/test_model_flavors.py:22]: Just do `2e1` or `20` here, unless there's some reason this multiplication needs to happen at runtime

File: test/pythonml/test_model_flavors.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            elif category == 'classification':
+                confusion_matrix = context.load('confusion_matrix')
+                accuracy = confusion_matrix.diagonal() / confusion_matrix.sum(axis=0)
+                assert (accuracy.all() >= 0.6)
+
+    def check_pass_conditions_predicting(self):
+        assert (os.path.isfile('predictions.csv'))
```
Comment (AcylSilane) [inline test/pythonml/test_model_flavors.py:29]: Assert isn't a function; it doesn't need parenthesis. I'd recommend removing the parentheses from the assert statements. For example, `assert os.path.isfile('predictions.csv')` or something here

File: test/pythonml/test_model_flavors.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    def run_flavor_test(self, category, flavor, do_predict):
```
Comment (AcylSilane) [inline test/pythonml/test_model_flavors.py:31]: Generally, our convention is to have booleans prefixed by `is` when possible, such as `is_predicting` or something here. Helps with readability to stay consistent with that

File: test/pythonml/test_pre_processing_flavors.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    train_descriptors = settings.context.load("train_descriptors")
+    test_target = settings.context.load("test_target")
+    test_descriptors = settings.context.load("test_descriptors")
+    return train_target, train_descriptors, test_target, test_descriptors
+
+
+class TestPreProcessingScalerFlavors(BaseUnitTest):
+    """
+    This class performs unittests for the pre_processing scaler flavors.
+    We may only check regression data here, and assume it would work for the other data categories.
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:29]: We should also do this for the other categories, since these units do behave differently depending on the type of problem. Not necessarily as part of this code review, but I'd make a task in Jira to finish this up for clustering / classification

File: test/pythonml/test_pre_processing_flavors.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    """
+
+    def min_max_pass_condition(self, data):
+        for col in data.T:
+            self.assertAlmostEqual(1.0, np.amax(col))
+            self.assertAlmostEqual(0.0, np.amin(col))
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:35]: If we're going to have different pass conditions and tests for different scalers, why not just make a separate class for each scaler? Unless there's a good reason these have to be clumped in the same object.

File: test/pythonml/test_pre_processing_flavors.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        ['regression', "pyml:pre_processing:min_max_scaler:sklearn.pyi"],
+        ['regression', "pyml:pre_processing:standardization:sklearn.pyi"],
+    ]
+    @parameterized.expand(params)
+    def test_flavor(self, category, flavor):
+        self.custom_setup(category)
+        self.run_scaler_flavor_test(category, flavor)
+
+
+class TestPreProcessingDroppers(BaseUnitTest):
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:67]: Rename this to be closer to the flavor this is actually testing, e.g. `TestPreProcessingRemoveData` or something.

File: test/pythonml/test_pre_processing_flavors.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    """
+    This class performs unittests for the pre_processing dropper flavors.
+    """
+
+    def remove_duplicates_pass_condition(self, data):
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:72]: Similar comment as the preprocessor file - if the tests have different methods to both run them, and to check their results, that's a sign that it might be better to split them out into separate classes

File: test/pythonml/test_pre_processing_flavors.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    params = [
+        ['regression', "pyml:pre_processing:remove_duplicates:pandas.pyi"],
+        ['regression', "pyml:pre_processing:remove_missing:pandas.pyi"]
+    ]
+    @parameterized.expand(params)
+    def test_flavor(self, category, flavor):
+
+        self.custom_setup(category)
+
+        import settings
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:95]: Why import settings here?

File: test/pythonml/test_pre_processing_flavors.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+        # duplicate the last row
+        elif 'remove_duplicates' in flavor:
+            data = data.append(data.tail(1), ignore_index=True)
+
+        # From the fake data, get the target col
+        data_target = data.iloc[:, -1]
+        data_descriptors = data.iloc[:, :-1]
+
+        import settings
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:111]: Nothing happens on this line if the module is already imported.

File: test/pythonml/test_pre_processing_flavors.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    ]
+    @parameterized.expand(params)
+    def test_flavor(self, category, flavor):
+
+        self.custom_setup(category)
+
+        import settings
+        data = pandas.read_csv(settings.datafile)
+
+        # add in some np.nan to the data
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:98]: It might be better to just make this data a new fixture that we import, instead of relying on the other datafile not changing.

File: test/pythonml/test_pre_processing_flavors.py [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+    def remove_duplicates_pass_condition(self, data):
+        data = pandas.DataFrame(data)
+        self.assertFalse(data.duplicated().any())
+
+    def remove_missing_pass_condition(self, data):
+        data = pandas.DataFrame(data)
+        self.assertFalse(data.isnull().values.any())
+
+    def droppers_pass_conditions(self, data, flavor):
```
Comment (AcylSilane) [inline test/pythonml/test_pre_processing_flavors.py:80]: To keep in-line with the naming convention of unittest, we should name this `assert_droppers_path_conditions` or something. Do this for the other checks that were written in this PR.

**Review by AcylSilane — CHANGES_REQUESTED** [review]
Overall, good job on this. I like that you separated out the unit-tests into different classes and used inheritance for them. There are also some changes to address to make this better.  Also, try to move that base unit test class out of the fixtures, it doesn't really belong there.

Pycharm has some tools to help with some of the refactoring (e.g. renaming, extracting variables, etc) - https://www.jetbrains.com/help/pycharm/refactoring-source-code.html

File: test/pythonml/test_context.py (hunk 3 above)
Comment (bsmith24) [reply test/pythonml/test_context.py:17]: Decided to leave this as having both, if that's ok

File: test/pythonml/test_context.py (hunk 4 above)
Comment (bsmith24) [reply test/pythonml/test_context.py:30]: Added a description

File: test/pythonml/test_context.py (hunk 2 above)
Comment (bsmith24) [reply test/pythonml/test_context.py]: Done

File: test/pythonml/test_pre_processing_flavors.py (hunk 10 above)
Comment (bsmith24) [reply test/pythonml/test_pre_processing_flavors.py:30]: I'll make this a separate Jira task, to test Classification and Clustering scenarios. The scaler may have different behavior depending on problem category

File: test/pythonml/test_pre_processing_flavors.py (hunk 11 above)
Comment (bsmith24) [reply test/pythonml/test_pre_processing_flavors.py:43]: I found it easier to keep these here for now, to avoid having to make new files for inheritance to separate out the common parts.

File: test/pythonml/test_pre_processing_flavors.py (hunk 14 above)
Comment (bsmith24) [reply test/pythonml/test_pre_processing_flavors.py:140]: We have to re-import settings to here to have access to it

File: test/pythonml/test_pre_processing_flavors.py (hunk 16 above)
Comment (bsmith24) [reply test/pythonml/test_pre_processing_flavors.py:143]: I'll make a task for this. Thanks

File: test/pythonml/test_context.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        with open('settings.py', 'w') as file:
+            for line in settings_template:
+                line = re.sub("PROBLEM_CATEGORY_HERE", 'regression', line)
+                file.write(line)
+        import settings
+        self.context = settings.Context()
+
+    def tearDown(self):
+        for data in ['.job_context', 'settings.py']:
+            subprocess.call('rm -rf '+data, shell=True)
```
Comment (AcylSilane) [inline test/pythonml/test_context.py:26]: Unless we need actually need to spawn a new shell process, try to avoid using the shell `shell=True`

Could do `subprocess.call(['rm', '-rf', data])` instead

Also happens in the `unittest.baseclass` definition

**Review by AcylSilane — CHANGES_REQUESTED** [review]
One more minor change, otherwise LGTM

**Review by timurbazhirov — APPROVED** [review]

**Review by AcylSilane — APPROVED** [review]
image
